enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+them accordingly. The existing image pulling functionality should be refactored in order to promote
+code reusability and avoid code duplication.
+
+### Workflow Description
+
+The general workflow for pre-caching user specified images is as follows:
+`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> <TALM workflow continues>`.
+
+Detailed user and worklow steps are outlined below.
+
+1. User creates a CGU CR with the `preCaching` field set to `true` and prepares a ConfigMap (named
+`cluster-group-upgrade-overrides`) with the list of additional images to be pre-cached as shown in
+the examples below. Note that the ConfigMap should be created concurrently with the CGU or should
+exist prior to the creation of it. Furthermore, the ConfigMap should be created in the same namespace
+as the CGU on the hub cluster.
+   - Example CGU CR:
+    ```yaml
+    apiVersion: ran.openshift.io/v1alpha1
+    kind: ClusterGroupUpgrade
+    metadata:
+      name: cgu
+      namespace: default
+    spec:
+      clusters:
+      - sno1
+      - sno3
+      - sno7
+      enable: false
+      preCaching: true
+      backup: false
+      managedPolicies:
+      - policy0-common-config-policy
+      remediationStrategy:
+        maxConcurrency: 4
+        timeout: 240
+    ```
+   - Example ConfigMap:
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-group-upgrade-overrides
+      namespace: default
+    data:
+      additionalWorkloadImages: |
+        quay.io/foobar/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
+        quay.io/foobar/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
+        ...
+        quay.io/foobar/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
+    ```
+
+2. The CR is validated (a function verifies if the clusters specified in the CR are part of a `ManagedCluster`
+object).
+
+3. If the validation succeeds, the pre-caching reconciliation commences. Note that all the sites are
+pre-cached concurrently.
+
+4. The pre-cache `spec` is then validated by checking if the managed policies exist. A valid `spec` will
+result in the following condition:
+    ```text
+    Type:    "PrecacheSpecValid",
+    Status:  True,
+    Reason:  "PrecacheSpecIsWellFormed",
+    Message: "Precaching spec is valid and consistent"
+    ```
+
+5. Pre-caching commences upon successful validation of the pre-cache `spec`. The following conditions are
+set while pre-caching is in-progress:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  False,
+    Reason:  "InProgress",
+    Message: "Precaching is required and not done"
+    ```
+
+6. After the pre-caching is successfully completed, the following condition is set:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  True,
+    Reason:  "PrecachingCompleted",
+    Message: "Precaching is completed for all clusters"
+    ```
```

Comment:
if precaching fails for some of the images for any reason, can we see which images are failed and why? if any of the additional images are failed to be pulled do we fail the precaching stage?

Reply:
Images that were not successfully pulled are logged accordingly. The reason behind the failure would need to be further investigated by the user. We handle the status of the additional images in the same way as the other images that need to pre-cached.

Reply:
Probably an appropriate message will help the user to investigate further with the debug